packages/dds/cell/src/cell.ts

```diff
@@ -291,8 +310,45 @@ export class SharedCell<T = any> extends SharedObject<ISharedCellEvents<T>>
      */
     protected applyStashedOp(content: unknown): unknown {
         const cellContent = content as ICellOperation;
-        this.applyInnerOp(cellContent);
-        ++this.messageId;
-        return this.messageId;
+        const previousValue = this.applyInnerOp(cellContent);
+        return createLocalOpMetadata(cellContent, this.getMessageId(), previousValue);
+    }
+
+    /**
+     * Rollback a local op
+     * @param op - The operation to rollback
+     * @param localOpMetadata - The local metadata associated with the op.
+     */
+    protected rollback(op: any, localOpMetadata: ICellLocalOpMetadata) {
+        if (op.type === "setCell" || op.type === "deleteCell") {
+            if (localOpMetadata.previousValue === undefined) {
+                this.deleteCore();
+            } else {
+                this.setCore(localOpMetadata.previousValue);
+            }
+
+            const lastPendingMessageId = this.pendingMessageIds.pop();
+            if (lastPendingMessageId !== localOpMetadata.pendingMessageId) {
+                throw new Error("Rollback op does not match last pending");
+            }
+        } else {
+            throw new Error("Unsupported op for rollback");
+        }
+    }
+
+     private getMessageId(): number {
```

Comment:
Is there any reason not to combine this method with createLocalOpMetadata? Seems unnecessary to always call both.